utils/docs/gclean-symbol.1.in
.TH gclean-symbol 1 "@DATE@" "gEDA Project" @VERSION@

.SH NAME
gclean-symbol \- clean gEDA symbol

.SH SYNOPSIS
.B gclean-symbol
.I symbol_file

.SH DESCRIPTION
.B gclean-symbol
clean gEDA symbol

.B usage: gclean-symbol old_symbol_file [ROTATE]> beautyfied_symbol_file
         ROTATE will rotate the labels of top and bottom pins

.SH HOMEPAGE
http://www.geda-project.org

.SH BUGS
You can report bugs at the Launchpad site of gclean-symbol:
https://bugs.launchpad.net/geda/gclean-symbol/+filebug

.SH AUTHOR
Werner Hoch

.SH SEE ALSO
.BR gsymcheck (1)

.SH COPYRIGHT
.nf
Copyright \(co 2008-@YEAR@ Werner Hoch.  License GPLv2+: GNU GPL
version 2 or later.  Please see the "COPYING" file included with this
program for full details.
.PP
This is free software: you are free to change and redistribute it. There is NO
WARRANTY; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE, to
the extent permitted by law.
.


NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2017 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf gEDA 2.1.1.1 - 04/01/19 Rampage Edition

* Expanded support for internationalization (i18n).

* All utilities now support internationalization (i18n).

* The visibility of toolbar tooltips can now be control using menu options.

* The context menu available when creating Polygon objects now includes
  an UNDO option allowing users to undo the previous section and resume
  path mode from the previous node.

* Scripting to support printing schematic documents from the command-line
  using gschem has been enhanced to support postscript or pdf output.
  Previously the default was changed from postscript to pdf. Postscript
  outputs could still be produced using menu options available in gschem.
  The output type is now automatically detected when using the command
  line interface by the file extension of the output file name. The
  default remains pdf if the output name is not given.

* Configuration function output-type has been renamed output-type.

* Libgedacairo now uses round end-caps when drawing buses, resulting in
  smoother transitions and improved appearance.

* A new keyword option hide-columns has been added to gattrib to allow
  user to designate a list attributes to be hidden in the GUI be default.
  A new Column Visibility dialog has been added to support user management
  of visible columns.

* A new path backend has been added to the gaf tool to allow scripting
  lanauages other than scheme to retrieve path information from libgeda.

* Improved appearence of various dialogs in gschem.

* gschem now logs exiting of the schematic editor if log-time is enabled.

* gschem now reports failure to attach attributes that aleady attached.

* gschem text properties dialog now supports operations to manipulate
  text case.

* gschem translate dialog now allows for translation of only selected
  objects. Additionally, the translate offset entry now accepts X and
  Y coordinates.

* Auxiliary buffers can now be explicitly cleared using a right context
  menu on the corresponding paste menu item.

* The suite now include template documents install to the user's template
  folder so the user can create new schematic and symbol documents from
  the system's context menu.

* Issues resolved:

* Libgedacairo has been enhanced to resolve issues with the end points of
  arcs shifting at different zoom levels. The shifting sometimes resulted
  in misalignment to adjoining lines at higher magnification levels.

* Extensive revisions were made to the Rampage Edition to build the suite
  on MinGW platforms. The Rampage Edition is now fully functional under
  native Win32 host systems including i18n support.

* The Rename Session dialog now fills in the previous session name,
  previously to dialog presented the user with an empty name field.

* The locked bit is not preserved when copying components in gschem.

* gschem now restarts text search when users change text in the Find Text
  dialog and click "find" or hits ENTER instead of closing the dialog.

* gschem now raises the main window when interacting with the Array Dialog.
  Previously, user could selected other application windows and return to
  the dialog and select the button that lowered the dialog but did not
  ensure the main window was in the foreground.

* An issue was resolved with controls not being initially enabled on the
  Attributes tab of the Preferences Dialog if the Attributes tab was the
  "saved" initial tab in gschem. Switching tabs or the view list controls
  re-enabled the disabled controls. The saved tab is now restored after
  settings are uploaded to the Preferences Dialog and this is reflected
  in the initial state of controls on the Attributes tab.

* An issue was resolved with resetting component-dialog-attributes when
  a second list was encountered in rc files. If a second attributes list
  was processed gschem would fault when the list was accessed or exiting
  gschem.

* The drc2 back end no longer reports duplicate slots for on components
  that do not have slots. This generally occurred when components have
  duplicate reference designators because these would typically show up
  using the same slot number.

* Libgeda no longer reports an oddity regarding symbol versions if the
  oddity was due to the symbol file being missing.

* The scheme file geda-deprecated-config.scm is no longer installed.

Notable changes in gEDA/gaf gEDA 2.1.1.1 - 11/12/17 Rampage Edition

* Expanded support for internationalization (i18n).

* The build system now support using the --with-included-gettext option to
  configure the suite to build and use an in-source international library.

* Font-sizes for widgets on GedaToolbars can now be specified in RC files.

* The toolbar layout in gschem has been improved to better support labels.

* The visibility of toolbar tooltips can now be control using RC keywords.

* The Slot Editor Dialog now incorporates a spinner control to adjust the
  slot number of slotted components. The range is restricted to valid slots
  for the selected component.

* New bitmaps have been added;

      * geda-deslect.png
      * geda-find-attribute
      * geda-properties.png
      * geda-trash-can.png

* libgedathon no longer over-writes existing gafrc when the declare_local_sym
  feature is used.

* libgedathon now supports internationalization (i18n).

Notable changes in gEDA/gaf gEDA 2.1.1.0 - 01/01/17 Rampage Edition

* Improvements were made to both the build system and source code to ensure
  gEDA-GAF builds more reliably on 64 bit platforms. Testing was performed
  using Ubuntu 12.04 and Ubuntu 14.04 64 bit host systems.


utils/scripts/gclean-symbol.py
#!/usr/bin/python

# Copyright (C) 2008-2014 Werner Hoch <werner.ho@gmx.de>
# Released under the terms of the GNU General Public License, version 2

# usage: gclean-symbol old_symbol_file [ROTATE]> beautyfied_symbol_file
#        ROTATE will rotate the labels of top and bottom pins

# missing features:
# * multiline text is not supported yet
# * getopt cli

# the script can be wrapped with a shellscript to improve the symbol repair workflow
shellscript = """
#!/bin/sh
# script to make a better symbol repair work flow
# it's neccesary that you have a temp directory r/ for the repaired symbols
# and gclean-symbol have to be in $PATH
# I use a symlink s/ to the original symbol path for short commands.

sym_name=$(basename "$1")

# check the old symbol
gsymcheck -vv "$1"

# show me the old symbol
gschem "$1" & >/dev/null

# let the script create a new skeleton of the symbol (add missing attributes)
./gclean-symbol "$1" "$2" >"r/""$sym_name"

# now edit the new sceleton (it's faster using xemacs than using gschem)
xemacs "r/""$sym_name"

# after editing the symbol look at the new symbol and do refinement
gschem "r/""$sym_name" >/dev/null

# after leaving gschem, check the new symbol
gsymcheck -vv "r/$sym_name"

# print a string to stdout that can be used to overwrite the old symbol
# with copy and paste.
# If the symbol is still bad, it can be edited with gschem or editor again.
echo
echo "   mv r/""$sym_name"" ""$1"
echo
"""

import re,sys

################# constants ########################################
VERSION="0.0.2"

PINTYPE=['in','out','io','oc','oe','pas','tp','tri','clk','pwr']

################ functions #####################################################
def makenice(filename, outfile):
    f=open(filename)
    file=[l.strip() for l in f.readlines()]
    pins=[]
    attr={}
    graphical_text=[]
    version=""
    nr =len(file)

    ## check for a valid symbol file
    ## only newer version are supported
    ##     if re.match("^v [0-9]{8} 1$",file[0]):
    ##        outfile.write("v 20070526 1\n")
    ##     else:
    ##        sys.stderr.write("ERROR: line 1: only newer file format versions are supported\n")

    i = 1
    while i < nr:
        if file[i][0] == 'T':
            if len(file[i+1].split("=")) == 2: ## attribute text
                t=file[i].split()
                if re.match('^refdes=',file[i+1]):
                    outfile.write(" ".join(t[0:3]) + ' 8 10 1 1 ' + " ".join(t[7:]) + "\n")
                    attr["refdes"]=1
                else:
                    outfile.write(" ".join(t[0:3]) + ' 5 10 0 0 ' + " ".join(t[7:]) + "\n")
                    attr[file[i+1].split("=")[0]] = 1
                outfile.write(file[i+1] + "\n")
            else:
                ## collect all graphical text elements and move them to the end
                ## of the file. Thus deleting is much easier with a text editor
                graphical_text.extend(file[i:i+2])
            i=i+2
            continue
        elif file[i][0] == "P":
            pin=["","","","",""]   ## Pin[pin,pinnumber,pinseq,pinlabel,pintype]
            pin[0]=file[i]
            outfile.write(pin[0] + "\n")
            i = i + 1
            if file[i][0] != "{":
                sys.stderr.write("ERROR: line %d: pin has no attributes\n" %i)
                escape=1
            else:
                escape=0
                i = i + 1
            outfile.write("{" + "\n")
            x,y,pos = getpinpos(pin[0])
            while file[i][0] != '}' and escape == 0:
                if re.match("^pinnumber=",file[i+1]) and file[i][0] == 'T':
                    if pos == "left":
                        pin[1] = 'T %i' %(x+200) + ' %i' %(y+50) + ' 5 8 1 1 0 6 1'
                    elif pos == "right":
                        pin[1] = 'T %i' %(x-200) + ' %i' %(y+50) + ' 5 8 1 1 0 0 1'
                    elif pos == "top":
                        pin[1] = 'T %i' %(x+50) + ' %i' %(y-200) + ' 5 8 1 1 0 0 1'
                        if ROTATE == 1:
                            pin[1] = 'T %i' %(x-50) + ' %i' %(y-200) + ' 5 8 1 1 90 0 1'
                    elif pos == "bottom":
                        pin[1] = 'T %i' %(x+50) + ' %i' %(y+100) + ' 5 8 1 1 0 0 1'
                        if ROTATE == 1:
                            pin[1] = 'T %i' %(x-50) + ' %i' %(y+200) + ' 5 8 1 1 90 6 1'
                    outfile.write(pin[1] + "\n" + file[i+1] + "\n")
                elif re.match("^pinseq=",file[i+1]) and file[i][0] == 'T':
                    if pos == "left":
                        pin[2] = 'T %i' %(x+200) + ' %i' %(y-50) + ' 5 8 0 1 0 8 1'
                    elif pos == "right":
                        pin[2] = 'T %i' %(x-200) + ' %i' %(y-50) + ' 5 8 0 1 0 2 1'
                    elif pos == "top":
                        pin[2] = 'T %i' %(x+50) + ' %i' %(y-200) + ' 5 8 0 1 0 2 1'
                        if ROTATE == 1:
                            pin[2] = 'T %i' %(x+50) + ' %i' %(y-200) + ' 5 8 0 1 90 2 1'
                    elif pos == "bottom":
                        pin[2] = 'T %i' %(x+50) + ' %i' %(y+100) + ' 5 8 0 1 0 2 1'
                        if ROTATE == 1:
                            pin[2] = 'T %i' %(x+50) + ' %i' %(y+200) + ' 5 8 0 1 90 8 1'
                    outfile.write(pin[2] + "\n" + file[i+1] + "\n")
                elif re.match("^pinlabel=",file[i+1]) and file[i][0] == 'T':
                    if pos == "left":
                        pin[3] = 'T %i' %(x+350) + ' %i' %y + ' 9 8 1 1 0 0 1'
                    elif pos == "right":
                        pin[3] = 'T %i' %(x-350) + ' %i' %y + ' 9 8 1 1 0 6 1'
                    elif pos == "top":
                        pin[3] = 'T %i' %x + ' %i' %(y-350) + ' 9 8 1 1 0 5 1'
                        if ROTATE == 1:
                            pin[3] = 'T %i' %x + ' %i' %(y-350) + ' 9 8 1 1 90 6 1'
                    elif pos == "bottom":
                        pin[3] = 'T %i' %x + ' %i' %(y+350) + ' 9 8 1 1 0 3 1'
                        if ROTATE == 1:
                            pin[3] = 'T %i' %x + ' %i' %(y+350) + ' 9 8 1 1 90 0 1'
                    outfile.write(pin[3] + "\n" + file[i+1] + "\n")
                elif re.match("^pintype=",file[i+1]) and file[i][0] == 'T':
                    if pos == "left":
                        pin[4] = 'T %i' %(x+350) + ' %i' %y + ' 5 8 0 1 0 2 1'
                    elif pos == "right":
                        pin[4] = 'T %i' %(x-350) + ' %i' %y + ' 5 8 0 1 0 8 1'
                    elif pos == "top":
                        pin[4] = 'T %i' %x + ' %i' %(y-500) + ' 5 8 0 1 0 5 1'
                        if ROTATE == 1:
                            pin[4] = 'T %i' %x + ' %i' %(y-350) + ' 5 8 0 1 90 8 1'
                    elif pos == "bottom":
                        pin[4] = 'T %i' %x + ' %i' %(y+500) + ' 5 8 0 1 0 3 1'
                        if ROTATE == 1:
                            pin[4] = 'T %i' %x + ' %i' %(y+350) + ' 5 8 0 1 90 2 1'
                    outfile.write(pin[4] + "\n" + file[i+1] + "\n")
                else:
                    sys.stderr.write("ERROR: line %d: pinsection wrong\n" %i)
                i = i + 2
            if pin[1] == "":
                if pos == "left":
                    pin[1] = 'T %i' %(x+200) + ' %i' %(y+50) + ' 5 8 1 1 0 6 1'
                elif pos == "right":
                    pin[1] = 'T %i' %(x-200) + ' %i' %(y+50) + ' 5 8 1 1 0 0 1'
                elif pos == "top":
                    pin[1] = 'T %i' %(x+50) + ' %i' %(y-200) + ' 5 8 1 1 0 0 1'
                    if ROTATE == 1:
                        pin[1] = 'T %i' %(x-50) + ' %i' %(y-200) + ' 5 8 1 1 90 0 1'
                elif pos == "bottom":
                    pin[1] = 'T %i' %(x+50) + ' %i' %(y+100) + ' 5 8 1 1 0 0 1'
                    if ROTATE == 1:
                        pin[1] = 'T %i' %(x-50) + ' %i' %(y+200) + ' 5 8 1 1 90 6 1'
                outfile.write(pin[1] + "\n" + "pinnumber=X" + "\n")
            if pin[2] == "":
                if pos == "left":
                    pin[2] = 'T %i' %(x+200) + ' %i' %(y-50) + ' 5 8 0 1 0 8 1'
                elif pos == "right":
                    pin[2] = 'T %i' %(x-200) + ' %i' %(y-50) + ' 5 8 0 1 0 2 1'
                elif pos == "top":
                    pin[2] = 'T %i' %(x+50) + ' %i' %(y-200) + ' 5 8 0 1 0 2 1'
                    if ROTATE == 1:
                        pin[2] = 'T %i' %(x+50) + ' %i' %(y-200) + ' 5 8 0 1 90 2 1'
                elif pos == "bottom":
                    pin[2] = 'T %i' %(x+50) + ' %i' %(y+100) + ' 5 8 0 1 0 2 1'
                    if ROTATE == 1:
                        pin[2] = 'T %i' %(x+50) + ' %i' %(y+200) + ' 5 8 0 1 90 8 1'
                outfile.write(pin[2] + "\n" + "pinseq=X" + "\n")
            if pin[3] == "":
                if pos == "left":
                    pin[3] = 'T %i' %(x+350) + ' %i' %y + ' 9 8 1 1 0 0 1'
                elif pos == "right":
                    pin[3] = 'T %i' %(x-350) + ' %i' %y + ' 9 8 1 1 0 6 1'
                elif pos == "top":
                    pin[3] = 'T %i' %x + ' %i' %(y-350) + ' 9 8 1 1 0 5 1'
                    if ROTATE == 1:
                        pin[3] = 'T %i' %x + ' %i' %(y-350) + ' 9 8 1 1 90 6 1'
                elif pos == "bottom":
                    pin[3] = 'T %i' %x + ' %i' %(y+350) + ' 9 8 1 1 0 3 1'
                    if ROTATE == 1:
                        pin[3] = 'T %i' %x + ' %i' %(y+350) + ' 9 8 1 1 90 0 1'
                outfile.write(pin[3] + "\n" + "pinlabel=X" + "\n")
            if pin[4] == "":
                if pos == "left":
                    pin[4] = 'T %i' %(x+350) + ' %i' %y + ' 5 8 0 1 0 2 1'
                elif pos == "right":
                    pin[4] = 'T %i' %(x-350) + ' %i' %y + ' 5 8 0 1 0 8 1'
                elif pos == "top":
                    pin[4] = 'T %i' %x + ' %i' %(y-500) + ' 5 8 0 1 0 5 1'
                    if ROTATE == 1:
                        pin[4] = 'T %i' %x + ' %i' %(y-350) + ' 5 8 0 1 90 8 1'
                elif pos == "bottom":
                    pin[4] = 'T %i' %x + ' %i' %(y+500) + ' 5 8 0 1 0 3 1'
                    if ROTATE == 1:
                        pin[4] = 'T %i' %x + ' %i' %(y+350) + ' 5 8 0 1 90 2 1'
                outfile.write(pin[4] + "\n" + "pintype=X" + "\n")
            if file[i][0] == "}":
                i = i + 1
            outfile.write("}" + "\n")

        elif file[i][0] in "ABLV":
            outfile.write(file[i] +"\n")
            i = i+ 1
        else:
            sys.stderr.write("ERROR: line %d: unknown file format error\n" % i)
            i = i + 1

    ## add attribute templates of missing attributes
    ay = 250
    for newattr in ["footprint", "numslots", "symversion", "use-license",
                    "dist-license", "author", "description", "documentation"]:
        if not attr.has_key(newattr):
            outfile.write("T 4000 %d 5 10 0 0 0 0 1" % ay + "\n")
            outfile.write(newattr + "=X" + "\n")
            ay += 200

    ## print all graphical text at the end of file
    outfile.write("\n".join(graphical_text))
    outfile.write("\n")

def getpinpos(pinstring):
    p = pinstring.split(" ")
    whichend = int(p[7])
    x1=int(p[1])
    x2=int(p[3])
    y1=int(p[2])
    y2=int(p[4])

    if whichend == 1:
        x1, y1, x2, y2 = x2, y2, x1, y1

    if x1 == x2: ## top or bottom pin
        if y1 < y2:
            pos = "bottom"
        else:
            pos = "top"
    else:  ## left or right pin
        if x1 < x2:
            pos ="left"
        else:
            pos = "right"

    return x1, y1, pos

########################## MAIN ################################################

try:
    file_in=sys.argv[1]
except:
    print "gclean-symbol:"
    print "Bad arguments, usage is: ", sys.argv[0] ,"infile [ROTATE]"
    sys.exit()

try:
    if sys.argv[2] == "ROTATE":
        ROTATE=1
    else:
        ROTATE=0
except:
    ROTATE=0

makenice(file_in, sys.stdout)
